--- a/riscv/riscv_vector_instruction_helpers.h
+++ b/riscv/riscv_vector_instruction_helpers.h
@@ -673,9 +673,12 @@
   }
   int sew = rv_vector->selected_element_width();
   int lmul = rv_vector->vector_length_multiplier();
-  int lmul_vd = lmul * sizeof(Vd) / sew;
-  int lmul_vs2 = lmul * sizeof(Vs2) / sew;
-  int lmul_vs1 = lmul * sizeof(Vs1) / sew;
+  // In vector reductions, the destination (vd) and scalar source (vs1) are
+  // always a single vector register group, effectively LMUL=1, so we use 8 (the
+  // base multiplier).
+  int lmul_vd = 8 * static_cast<int>(sizeof(Vd)) / sew;
+  int lmul_vs2 = lmul * static_cast<int>(sizeof(Vs2)) / sew;
+  int lmul_vs1 = 8 * static_cast<int>(sizeof(Vs1)) / sew;
   if (lmul_vd > 64 || lmul_vs2 > 64 || lmul_vs1 > 64) {
     rv_vector->set_vector_exception();
     LOG(ERROR) << "Illegal lmul value";
